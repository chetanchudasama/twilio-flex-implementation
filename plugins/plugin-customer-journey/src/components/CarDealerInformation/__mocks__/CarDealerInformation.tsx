import React from "react";
import { Props } from "../CarDealerInformation";

const CarDealerInformation: React.FC<Props> = (props) => <div />;

export default CarDealerInformation;
