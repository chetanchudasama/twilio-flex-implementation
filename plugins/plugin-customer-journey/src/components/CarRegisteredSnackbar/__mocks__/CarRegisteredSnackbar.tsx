import React from "react";
import { CarRegisteredSnackbarProps } from "../CarRegisteredSnackbar.Props";

export const CarRegisteredSnackbar: React.FC<CarRegisteredSnackbarProps> = (
  props
) => {
  return <div className="car-registered-snackbar">CarRegisteredSnackbar</div>;
};
