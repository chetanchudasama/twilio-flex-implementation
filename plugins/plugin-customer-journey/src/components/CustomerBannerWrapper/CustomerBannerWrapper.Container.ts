import {
  CustomerBannerProps,
  CustomerDetailModel,
  Shared,
} from "@common/components";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";
import { AppState, namespace } from "../../states";
import { Actions } from "../../states/customer";
import CustomerBannerWrapper from "./CustomerBannerWrapper";
import { HistoryDetailActions } from "../../states/historyDetail";
import { getWorkerIsAvailableForOutboundCall } from "../../states/selectors";

export interface StateToProps extends CustomerBannerProps {
  isCustomerAvailable: boolean;
  applicationId: number;
  canMakeCall: boolean;
}

export interface DispatchToProps {
  setCustomer: (customer: CustomerDetailModel) => void;
  setCustomerDetails: (key: keyof CustomerDetailModel, value: any) => void;
  addNoteDetail: (content: string, isImportant: boolean) => void;
}

const mapStateToProps = (state: AppState): StateToProps => {
  const customer: CustomerDetailModel = Object.assign(
    new CustomerDetailModel(),
    state[namespace].customer.customer
  );
  const name = customer.fullName;
  const dob = Shared.getFormattedDate(customer.dateOfBirth, "DD/MM/YYYY");
  const {
    address,
    mobileNumber,
    emailAddress,
    callbackBooked,
    applicationId,
    hasThirdPartyAuthorization,
    thirdPartyAuthorization,
    hasVulnerableCustomerReported,
    vulnerableCustomerInformation,
  } = customer;

  const canMakeCall = getWorkerIsAvailableForOutboundCall(state);
  const { applicationStatusName } = customer.applicationStatus;
  const { lenderName, tierName } = customer.preferredLender;
  const apr = customer.preferredLender.apr ?? 0;
  return {
    isCustomerAvailable: !!state[namespace].customer.customer,
    name,
    dob,
    address,
    mobileNumber,
    emailAddress,
    callbackBooked,
    applicationId,
    hasThirdPartyAuthorization,
    applicationStatusName,
    hasVulnerableCustomerReported,
    thirdPartyAuthorization,
    vulnerableCustomerInformation,
    saveThirdPartyDetailHandler: () => {},
    saveVulnerableCustomerInformation: () => {},
    setCallbackDetail: () => {},
    canMakeCall,
    makeCall: () => {},
    lenderName,
    apr,
    tierName,
    updateCustomerDetail: () => {},
  };
};

const mapDispatchToProps = (dispatch: Dispatch<any>): DispatchToProps => ({
  setCustomer: bindActionCreators(Actions.setCustomer, dispatch),
  setCustomerDetails: bindActionCreators(Actions.setCustomerDetails, dispatch),
  addNoteDetail: bindActionCreators(
    HistoryDetailActions.addNoteDetail,
    dispatch
  ),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CustomerBannerWrapper);
