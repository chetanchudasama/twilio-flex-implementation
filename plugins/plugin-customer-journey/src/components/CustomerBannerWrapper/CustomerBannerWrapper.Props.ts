import { CustomerBannerProps, CustomerDetailModel } from "@common/components";

export interface StateToProps extends CustomerBannerProps {
  isCustomerAvailable: boolean;
  applicationId: number;
  canMakeCall: boolean;
}

export interface DispatchToProps {
  setCustomer: (customer: CustomerDetailModel) => void;
  setCustomerDetails: (key: keyof CustomerDetailModel, value: any) => void;
  addNoteDetail: (content: string, isImportant: boolean) => void;
}
