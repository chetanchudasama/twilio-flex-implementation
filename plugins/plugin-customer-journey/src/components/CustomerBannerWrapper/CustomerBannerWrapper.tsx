import {
  CustomerBanner,
  ThirdPartyAuthorizationDetail,
  VulnerableCustomerInformation,
  Shared,
  CustomerDetailType,
} from "@common/components";
import { Actions, Manager } from "@twilio/flex-ui";
import React from "react";
import moment from "moment";
import {
  CustomNotificationType,
  showErrorMessage,
  showMessage,
} from "../../Notifications";
import { useApplicationService } from "../../services/application.service";
import { AppState } from "../../states";
import { CustomerCallbackRequestModel } from "../../models/CustomerCallbackRequestModel";
import { UpdateDetailRequestModel } from "../../models/UpdateDetailRequestModel";
import type {
  DispatchToProps,
  StateToProps,
} from "./CustomerBannerWrapper.Props";

type Props = StateToProps & DispatchToProps;

const CustomerBannerWrapper: React.FC<Props> = (props) => {
  const {
    applicationId,
    isCustomerAvailable,
    canMakeCall,
    mobileNumber,
    setCustomerDetails,
    addNoteDetail,
  } = props;
  const state: AppState = Manager.getInstance().store.getState();
  const applicationService = useApplicationService(
    state.flex.session.ssoTokenPayload.token
  );

  const makeCallHandler = () => {
    if (canMakeCall && mobileNumber) {
      const payload = {
        destination: props.mobileNumber,
      };
      Actions.invokeAction("StartOutboundCall", payload);
    }
  };

  const saveThirdPartyDetailHandlerFunction = async (
    thirdPartyAuthorizationInfo: ThirdPartyAuthorizationDetail | null
  ) => {
    const updateRequestList: UpdateDetailRequestModel[] = [
      {
        op: "replace",
        path: "/thirdPartyAuthorization",
        value: thirdPartyAuthorizationInfo,
      },
      {
        op: "replace",
        path: "/hasThirdPartyAuthorization",
        value: !!thirdPartyAuthorizationInfo,
      },
    ];

    try {
      await applicationService.updateBaseApplicationDetail(
        applicationId,
        updateRequestList
      );
      setCustomerDetails(
        "thirdPartyAuthorization",
        thirdPartyAuthorizationInfo
      );
      setCustomerDetails(
        "hasThirdPartyAuthorization",
        !!thirdPartyAuthorizationInfo
      );
      // show success message
      showMessage(
        CustomNotificationType.SuccessNotification,
        "Third party authorization detail updated successfully!"
      );
    } catch (error) {
      showErrorMessage("Something went wrong, please try again!", "", true);
    }
  };

  const saveVulnerabilityHandler = async (
    vulnerableCustomerDetail: VulnerableCustomerInformation | null
  ) => {
    const updateRequestList: UpdateDetailRequestModel[] = [
      {
        op: "replace",
        path: "/vulnerableCustomerInformation",
        value: vulnerableCustomerDetail,
      },
      {
        op: "replace",
        path: "/hasVulnerableCustomerReported",
        value: !!vulnerableCustomerDetail,
      },
    ];
    try {
      await applicationService.updateBaseApplicationDetail(
        props.applicationId,
        updateRequestList
      );
      setCustomerDetails(
        "hasVulnerableCustomerReported",
        !!vulnerableCustomerDetail
      );
      setCustomerDetails(
        "vulnerableCustomerInformation",
        vulnerableCustomerDetail
      );
      // show success message
      showMessage(
        CustomNotificationType.SuccessNotification,
        "Customer vulnerability detail updated successfully!"
      );
    } catch (error) {
      showErrorMessage("Something went wrong, please try again!", "", true);
    }
  };

  const handleSetCallbackDetail = async (
    callbackDate: Date | null,
    reason = ""
  ) => {
    if (callbackDate) {
      // submit callback detail
      const callbackDetailModel: CustomerCallbackRequestModel =
        new CustomerCallbackRequestModel();
      callbackDetailModel.callbackDate = moment(callbackDate).utc();
      callbackDetailModel.note = reason;
      try {
        await applicationService.submitCallbackDetail(
          applicationId,
          callbackDetailModel
        );
        setCustomerDetails("callbackBooked", callbackDate);
        addNoteDetail(
          `Callback booked (${Shared.getFormattedDate(
            callbackDate,
            "DD/MM/YYYY HH:mm"
          )})\n${reason}`,
          false
        );
      } catch (error) {
        showErrorMessage("Something went wrong, please try again!", "", true);
      }
    } else {
      // delete callback detail
      const updateRequestList = [
        {
          op: "replace",
          path: "/callbackBooked",
          value: null,
        },
      ];
      try {
        await applicationService.updateBaseApplicationDetail(
          applicationId,
          updateRequestList
        );
        setCustomerDetails("callbackBooked", null);
      } catch (error) {
        showErrorMessage("Something went wrong, please try again!", "", true);
      }
    }
  };

  const updateCustomerDetail = (type: CustomerDetailType) => {
    Actions.invokeAction("updateCustomerDetail", {
      customerDetailType: type,
    });
  };

  const banner = (
    <CustomerBanner
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}
      saveThirdPartyDetailHandler={saveThirdPartyDetailHandlerFunction}
      saveVulnerableCustomerInformation={saveVulnerabilityHandler}
      setCallbackDetail={handleSetCallbackDetail}
      makeCall={makeCallHandler}
      updateCustomerDetail={updateCustomerDetail}
    />
  );

  return <>{isCustomerAvailable ? banner : null}</>;
};

export default CustomerBannerWrapper;
