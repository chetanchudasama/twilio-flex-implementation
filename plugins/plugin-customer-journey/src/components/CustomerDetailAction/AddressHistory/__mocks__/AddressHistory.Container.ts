import AddressHistory from "./AddressHistory";

export default AddressHistory;
