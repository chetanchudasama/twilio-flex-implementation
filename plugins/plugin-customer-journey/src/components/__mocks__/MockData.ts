import { WorkerAttributes } from "@twilio/flex-ui";
import { namespace } from "../../states";

const workerAttributes: WorkerAttributes = {
  contact_uri: "client:test",
  email: "test@test.com",
  full_name: "test user",
  image_url: "img_url",
  roles: ["admin"],
  agentId: 1,
};

const conferenceSource = {
  source: {
    participants: [
      {
        callSid: "callSid",
        participantType: "customer",
        workerSid: "",
      },
      {
        callSid: "callSid",
        participantType: "worker",
        workerSid: "workerSid",
      },
    ],
  },
};

const conferenceState = new Map();
conferenceState.set("taskSid", conferenceSource);

export const flexManagerMock = jest.fn().mockReturnValue({
  workerClient: {
    source: {},
    sid: "workerSid",
    name: "test user",
    fullName: "test user",
    activityName: "Offline",
    attributes: workerAttributes,
    dateUpdated: new Date(),
    isAvailable: false,
    activityDuration: "",
    dateStatusChanged: new Date(1).toISOString(),
    activity: {
      available: false,
      name: "Offline",
      sid: "OfflineSid",
    },
  },
  store: {
    getState: jest.fn().mockReturnValue({
      flex: {
        worker: {
          attributes: workerAttributes,
        },
        session: {
          ssoTokenPayload: {
            token: "ssoToken",
          },
        },
        conferences: {
          states: conferenceState,
        },
      },
      [namespace]: {
        phoneNumberList: {
          phoneNumberList: [
            {
              friendlyName: "Sales",
              phoneNumber: "+1234567891",
              queueSid: "QUEUE_SID",
            },
          ],
        },
      },
    }),
  },
  strings: {
    DIALPADExternalTransferSearchInputLabel: "Search By Name or PhoneNumber",
    DIALPADExternalTransferHoverOver: "Add",
    DIALPADExternalTransferPopupHeader: "External Transfer",
  },
  serviceConfiguration: {
    outbound_call_flows: {
      default: {
        caller_id: "CALLER_ID",
      },
    },
  },
});
