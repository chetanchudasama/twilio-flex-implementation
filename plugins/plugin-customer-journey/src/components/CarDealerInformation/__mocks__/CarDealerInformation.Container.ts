import CarDealerInformation from "./CarDealerInformation";

export default CarDealerInformation;
