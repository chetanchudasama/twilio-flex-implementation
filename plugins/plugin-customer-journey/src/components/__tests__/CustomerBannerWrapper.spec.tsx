import { Manager } from "@twilio/flex-ui";
import { mount } from "enzyme";
import flushPromises from "flush-promises";
import React from "react";
import { create } from "react-test-renderer";
import { showMessage, showErrorMessage } from "../../Notifications";
import CustomerBannerWrapper from "../CustomerBannerWrapper/CustomerBannerWrapper";
import {
  DispatchToProps,
  StateToProps,
} from "../CustomerBannerWrapper/CustomerBannerWrapper.Container";
import { flexManagerMock } from "../__mocks__/MockData";

const props: StateToProps & DispatchToProps = {
  isCustomerAvailable: true,
  applicationId: 1,
  name: "Thomas Snow",
  dob: "",
  address: "Building1",
  mobileNumber: "+44567890",
  email: "thomas@gmail.com",
  callbackBooked: new Date(2021, 5, 21),
  setCustomer: () => jest.fn(() => {}),
  setCustomerDetails: () => jest.fn(() => {}),
  hasThirdPartyAuthorization: false,
  thirdPartyAuthorization: null,
  saveThirdPartyDetailHandler: () => jest.fn(() => {}),
  hasVulnerableCustomerReported: false,
  saveVulnerableCustomerInformation: jest.fn(),
  vulnerableCustomerInformation: null,
  setCallbackDetail: jest.fn(() => {}),
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  addNoteDetail: (_content: string, _isImportant: boolean) => null,
  canMakeCall: true,
  lenderName: "Auto loan",
  apr: 19.8,
  tierName: "Tier 1",
};

jest.mock("../../Notifications", () => {
  return {
    CustomNotificationType: { SuccessNotification: "successNotification" },
    showErrorMessage: jest.fn(() => {}),
    showMessage: jest.fn(() => {}),
  };
});

Manager.getInstance = flexManagerMock;

jest.mock("@material-ui/core", () => {
  const materialUI = jest.requireActual("@material-ui/core");
  return {
    ...materialUI,
    Tooltip: jest.fn(({ children }) => children),
  };
});

jest.mock("../../services/application.service");

const mockUpdateBaseApplicationDetail = jest.fn();
const mockSubmitCallbackDetail = jest.fn();

jest.mock("../../services/application.service", () => ({
  useApplicationService: jest.fn(() => ({
    updateBaseApplicationDetail: mockUpdateBaseApplicationDetail,
    submitCallbackDetail: mockSubmitCallbackDetail,
  })),
}));

jest.mock("@common/components", () => {
  const components = jest.requireActual("@common/components");
  return {
    ...components,
    CustomerBanner: jest.fn(
      ({
        children,
        saveThirdPartyDetailHandler,
        callbackBooked,
        setCallbackDetail,
      }) => {
        return (
          <div>
            Customer Banner
            <button
              type="button"
              className="save-third-party-detail"
              onClick={() => saveThirdPartyDetailHandler(null)}
            >
              Save Third Party Detail
            </button>
            <button
              type="button"
              className="set-callback"
              onClick={() => setCallbackDetail(callbackBooked, "test reason")}
            >
              Set Callback
            </button>
            <button
              type="button"
              className="delete-callback"
              onClick={() => setCallbackDetail(null, "")}
            >
              Delete
            </button>
            {children}
          </div>
        );
      }
    ),
  };
});

describe("CustomerBannerWrapper", () => {
  it("renders correctly", () => {
    const wrapper = create(<CustomerBannerWrapper {...props} />).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it("when value of customer prop is not passed, CustomerBanner component should not be rendered", async () => {
    props.isCustomerAvailable = false;
    const wrapper = mount(<CustomerBannerWrapper {...props} />);

    expect(wrapper).toEqual({});
  });

  it("when saveThirdPartyDetailHandler method is emitted from CustomerBanner component, updateBaseApplicationDetail API should be called", async () => {
    props.isCustomerAvailable = true;
    mockUpdateBaseApplicationDetail.mockImplementationOnce(() =>
      Promise.resolve(undefined)
    );
    const wrapper = mount(<CustomerBannerWrapper {...props} />);

    wrapper.find(".save-third-party-detail").simulate("click");

    await flushPromises();

    expect(showMessage).toHaveBeenCalledWith(
      "successNotification",
      "Third party authorization detail updated successfully!"
    );
  });

  it("when updateBaseApplicationDetail API returns an error, error message should be displayed", async () => {
    props.isCustomerAvailable = true;
    mockUpdateBaseApplicationDetail.mockImplementationOnce(() =>
      // eslint-disable-next-line prefer-promise-reject-errors
      Promise.reject({ statusCode: 500 })
    );
    const wrapper = mount(<CustomerBannerWrapper {...props} />);

    wrapper.find(".save-third-party-detail").simulate("click");

    await flushPromises();

    expect(showErrorMessage).toHaveBeenCalledWith(
      "Something went wrong, please try again!",
      "",
      true
    );
  });

  it("when set callback button is clicked from CustomerBanner component, submitCallbackDetail API should be called", async () => {
    mockSubmitCallbackDetail.mockImplementationOnce(() =>
      Promise.resolve(undefined)
    );
    const wrapper = mount(<CustomerBannerWrapper {...props} />);

    wrapper.find(".set-callback").simulate("click");

    await flushPromises();
    expect(mockSubmitCallbackDetail).toHaveBeenCalledTimes(1);
  });

  it("when submitCallbackDetail API returns an error, error message should be displayed", async () => {
    mockSubmitCallbackDetail.mockImplementationOnce(() =>
      // eslint-disable-next-line prefer-promise-reject-errors
      Promise.reject({ statusCode: 500 })
    );
    const wrapper = mount(<CustomerBannerWrapper {...props} />);

    wrapper.find(".set-callback").simulate("click");

    await flushPromises();

    expect(showErrorMessage).toHaveBeenCalledWith(
      "Something went wrong, please try again!",
      "",
      true
    );
  });

  it("when delete callback button is clicked from CustomerBanner component, updateBaseApplicationDetail API should be called", async () => {
    mockUpdateBaseApplicationDetail.mockImplementationOnce(() =>
      Promise.resolve(undefined)
    );
    const wrapper = mount(<CustomerBannerWrapper {...props} />);

    wrapper.find(".delete-callback").simulate("click");

    await flushPromises();
    expect(mockUpdateBaseApplicationDetail).toHaveBeenCalled();
  });
});
