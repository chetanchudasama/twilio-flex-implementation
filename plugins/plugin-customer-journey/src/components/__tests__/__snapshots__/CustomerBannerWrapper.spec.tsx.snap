// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerBannerWrapper renders correctly 1`] = `
<div>
  Customer Banner
  <button
    className="save-third-party-detail"
    onClick={[Function]}
    type="button"
  >
    Save Third Party Detail
  </button>
  <button
    className="set-callback"
    onClick={[Function]}
    type="button"
  >
    Set Callback
  </button>
  <button
    className="delete-callback"
    onClick={[Function]}
    type="button"
  >
    Delete
  </button>
</div>
`;
